File: src/physical/board.py
"""Module holding Board and Move classes.
All row values are 0-indexed,
so no matter what part of the engine you're in (unless it's user-facing),
row 7 is the equivalent of real-life row 8.

# Functions

## col_row(square: Square) -> tuple[str, int]
Returns the info of a square.

# Classes

## Square
Represents a square on the board.

## PieceColor
Enum for color of a player's pieces. Also used in other parts of the engine to distinguish players.

## PieceType
Enum for type of pieces, including value.

## Entity(typing.Protocol)
Only available in the `codon` branch of Sophisticate.
A protocol representing a piece on the board or lack thereof.

## NoPiece(Entity)
Only available in the `codon` branch of Sophisticate.
A little controdictory, but used instead of None on squares with no piece.

## Piece(Entity)
A piece on the board, holding color, type, and location.

## Board
The board itself. Holds columns, rows, and sets up starting position by default.
"""
from custom_enum import EnumMember, Enum
import copy
import time
import typing

#used to track how long duplication has taken
elapsed_duplication = 0.0


class Square:
    """# Square
    One square on the board.

    # Methods

    ## __init__(self, col: str, row: int)
    Col is a regular board column (a - h). Row is a zero-indexed row. So, what GothamChess would call H8 would be Square(\"h\", 7).

    ## draw(self) -> str
    Returns very bad ascii art of the square."""

    def __init__(self, col: str, row: int):
        self.col = col
        self.row = row
        self.piece: Piece | None = None

        #a very complicated-looking way to determine square color.
        #If the row number plus the column number are even, black. else white
        self.color = PieceColor.BLACK if (LETTERS.index(col) + row) % 2 == 0 else PieceColor.WHITE

    def __str__(self) -> str:
        return f"Square {self.col}{self.row} (0-indexed; {self.col}{self.row + 1} in standard notation)"

    def __repr__(self) -> str:
        return f"Square {self.col}{self.row} (0-indexed; {self.col}{self.row + 1} in standard notation)"

    def __eq__(self, other) -> bool:
        return self.piece == other.piece and self.col == other.col and self.row == other.row
    
    def __ne__(self, other) -> bool:
        return not self.__eq__(other)

    def __hash__(self) -> int:
        #be laaaaaaazy
        return hash(str(self))

    def draw(self) -> str:
        bg = "\033[40m" if self.color == PieceColor.BLACK else "\033[47m"
        fg = "\033[0;37m" if self.color == PieceColor.BLACK else "\033[30m"
        piece = "  " if self.piece == None else self.piece.two_letter()
        return f"{bg}{fg}{piece}\033[0m"


def col_row(square: Square) -> tuple[str, int]:
    return (square.col, square.row)


#a simple timer to track how long duplication takes
def duplication_clock(func):
    def timer(self):
        global elapsed_duplication
        start = time.time()
        result = func(self)
        elapsed_duplication += time.time() - start
        return result
    return timer


class PieceColor(Enum):
    """# PieceColor
    Enum for color of pieces.

    # Values

    BLACK = -1
    WHITE = 1
    """
    BLACK = -1
    WHITE = 1


class PieceType(Enum):
    """Enum representing types (Bishop, Queen, King, etc).

    # Values

    PAWN = 1
    KNIGHT = 3
    BISHOP = 4 because I personally like bishops more than knights
    ROOK = 5
    QUEEN = 9
    KING = 39 because all other  together are 39 and the king is worth the game.
    """

    NONE = 0
    PAWN = 1
    KNIGHT = 3
    BISHOP = 4
    ROOK = 5
    QUEEN = 9
    KING = 39


class Entity(typing.Protocol):
    """# Entity(typing.Protocol)
    A Protocol for a piece or lack thereof.

    # Methods

    ## __eq__(self, other), __ne__(self, other)
    Python dunder methods that must be implemented by the enheriting type.
    """

    def __eq__(self, other) -> bool:
        raise NotImplementedError

    def __ne__(self, other) -> bool:
        raise NotImplementedError


class NoPiece(Entity):
    """# NoPiece(Entity)
    A type used instead of Piece when there is no piece on a given square. Equivalent to None.
    No custom methods or attributes other than __eq__() and __ne__()
    """
    
    def __eq__(self, other) -> bool:
        return other == None or other is None

    def __ne__(self, other) -> bool:
        return not self == other

class Piece(Entity):
    """Class representing a on the 

    # Methods

    ## __init__(self, ptype: PieceType, color: PieceColor, location: Square)
    Pretty self-explanatory.

    ## two_letter(self) -> str
    Returns 2-letter string of piece: 'BQ' for black queen, 'WK' for white king, and so on.

    # Attributes

    ptype: PieceType - The type of the piece
    color: PieceColor - The color of the piece
    location: Square - The square the is on
    en_passant: bool=False - Whether the piece can be captured with en passant.
    """
    
    def __init__(self, ptype: EnumMember, color: PieceColor, location: Square):
        self.ptype = ptype
        self.color = color
        self.location = location
        self.has_moved = False
        self.en_passant = False

    def __str__(self) -> str:
        return f"{self.color.name} {self.ptype.name} on {self.location}"

    def __repr__(self) -> str:
        return f"{self.color.name} {self.ptype.name} on {self.location}"

    def __eq__(self, other) -> bool:
        if other == None: return False
        return self.ptype == other.ptype and self.color == other.color

    def __ne__(self, other) -> bool:
        return not self.__eq__(other)

    def __hash__(self) -> int:
        #laziness
        return hash(str(self))

    def two_letter(self) -> str:
        #                                                knights should be represented with n instead of k
        return f"{self.color.name[0]}{self.ptype.name[0] if self.ptype.name != 'KNIGHT' else 'N'}"


class Board:
    """Class to represent a chess board.
    Automatically populates with 8 lettered columns of 8 squares.

    # Attributes

    ## threatened_squares: list[Square]
    All squares on the board that can be moved to by any player.

    ## squares_white_threatens: list[Square]
    All squares on the board that can be moved to by white.

    ## squares_black_threatens: list[Square]
    All squares on the board that can be moved to by black.

    ## pieces: list[Piece]
    All pieces on the board.

    ## white_castled: bool=False
    Has white castled?

    ## black_castled: bool=False
    Has black castled?
    
    # Methods

    ## __init__(self, default_pos: bool=True)
    If default_pos is True, a stock starting position is automatically set up.

    ## __getitem__(self, col_name: str) -> list[Square]
    Takes a column name and returns a list of squares.

    ## white_pieces(self) -> list[Piece]
    Iterates through pieces and returns white pieces in no particular order.

    ## black_pieces(self) -> list[Piece]
    Iterates through pieces and returns white pieces in no particular order.

    ## set_up_game_board(self)
    Puts pieces in default positions for a standard game.

    ## prep_column(self, col_name: str, ptype: PieceType)
    Adds pawns and given piece in correct places to given column.

    ## add_pawns(self, col: list[Square])
    Adds pawns to second and second-to-last squares in column.

    ## reset_en_passant(self)
    Sets all pieces' en_pssant attributes to False.

    ## duplicate(self) -> Board
    Creates a new board identical to this one.

    ## clean(self)
    Similar logic to .duplicate(), but only acting on self and not producing a new board.
    Essentially the same as game = game.duplicate(), but acting on the object and without as much memory overhead.

    ## draw(self) -> str
    Draws an ascii representation of the board. The default for __str__ and __repr__.
    """

    def __init__(self, default_pos: bool=True):
        # Set up 8 * 8 board with dict to represent lettered columns
        self.squares: dict[str, list[Square]] = {}
        self.pieces = []
        self.threatened_squares = []
        self.squares_white_threatens = []
        self.squares_black_threatens = []
        self.white_castled, self.black_castled = False, False
        for col in "abcdefgh":
            col_list = []
            for row in range(8):
                col_list.append(Square(col, row))
            self.squares[col] = col_list

        if default_pos:
            self.set_up_game_board()

    def __getitem__(self, col_name: str) -> list[Square]:
        if not isinstance(col_name, str) or len(col_name) != 1 or col_name.isdecimal():
            raise TypeError("col_name is not a single-letter string.")
        return self.squares[col_name]

    def __str__(self) -> str:
        return self.draw()
    
    def __repr__(self) -> str:
        return self.draw()

    def __eq__(self, other) -> bool:
        for col_name in LETTERS:
            for row in range(8):
                if self[col_name][row] != other[col_name][row]: return False
        return True

    def __ne__(self, other) -> bool:
        return not self.__eq__(other)

    def black_pieces(self) -> list[Piece]:
        return [piece for piece in self.pieces if piece.color == PieceColor.BLACK]
    
    def white_pieces(self) -> list[Piece]:
        return [piece for piece in self.pieces if piece.color == PieceColor.WHITE]

    def set_up_game_board(self):
        self.prep_column("a", PieceType.ROOK)
        self.prep_column("h", PieceType.ROOK)
        self.prep_column("b", PieceType.KNIGHT)
        self.prep_column("g", PieceType.KNIGHT)
        self.prep_column("c", PieceType.BISHOP)
        self.prep_column("f", PieceType.BISHOP)
        self.prep_column("d", PieceType.QUEEN)
        self.prep_column("e", PieceType.KING)

    def prep_column(self, col_name: str, ptype: PieceType):
        col = self[col_name]
        col[0].piece = Piece(ptype, PieceColor.WHITE, col[0])
        col[7].piece = Piece(ptype, PieceColor.BLACK, col[7])
        self.pieces.append(col[0].piece)       
        self.pieces.append(col[7].piece)
        self.add_pawns(col)

    def add_pawns(self, col: list[Square]):
        col[1].piece = Piece(PieceType.PAWN, PieceColor.WHITE, col[1])
        col[6].piece = Piece(PieceType.PAWN, PieceColor.BLACK, col[6])
        self.pieces.append(col[1].piece)
        self.pieces.append(col[6].piece)

    def reset_en_passant(self):
        for piece in self.pieces:
            piece.en_passant = False
    
    #creates new identical board in O(ignorance is bliss) time.
    @duplication_clock
    def duplicate(self):
        new = Board(False)

        #iterate over columns and rows
        for col in LETTERS:
            for row in range(8):
                piece = self[col][row].piece
                if piece == None:
                    continue

                #copy piece
                new_piece = Piece(copy.deepcopy(piece.ptype), copy.deepcopy(piece.color), new[col][row])
                new.pieces.append(new_piece)
                new[col][row].piece = new_piece
        return new

    #considered duplication because the algorithm is almost identical
    @duplication_clock
    def clean(self):
        #erase ghosts not on board.
        self.pieces = []
        
        #iterate over cols and rows to find pieces
        for col in LETTERS:
            for row in range(8):
                piece = self[col][row].piece
                if piece != None: self.pieces.append(piece)

    def draw(self) -> str:
        output = "      a    b    c    d    e    f    g    h"
        #all row values, top to bottom
        for row in range(7, -1, -1):
            output += f"\n\n{row + 1}"
            for column in LETTERS:
                output += " | "
                square = self[column][row]
                output += square.draw()

        #Skip first 2 newlines
        return output[2::]


LETTERS = "abcdefgh"

File: src/physical/board.rs
//!# board
//!Holds peices, piece information, board, etc.
//!
//!As a reminder, all row numbers are 0-indexed; what GothamChess would call a8 is a7 here, and a1
//!is a0.

use crate::physical::movement::MoveInfo;
use std::fmt;
use std::mem::{self, MaybeUninit};
use std::collections::HashSet;

pub const LETTERS: &str = "abcdefgh";

///# PieceType
///Represents a piece type.
///
///# Public Methods
///
///## value(&self) -> i32
///Returns the value of a piece (1, 3, 5, 9, 50)
///
///## from(letter: char) -> Result<Self, ()>
///Creates Self from piece letter
///
///## letter(&self) -> char
///Gets a character representing the piece type.
#[derive(Debug, Eq, PartialEq, Clone, Copy)]
pub enum PieceType {
    Pawn(i32), //store turn pawn can be en passant'd on.
    Knight,
    Bishop,
    Rook,
    Queen,
    King,
}

impl PieceType {
    pub fn value(&self) -> i32 {
        match self {
            PieceType::Pawn(_) => 1,
            PieceType::Knight => 3,
            PieceType::Bishop => 3,
            PieceType::Rook => 5,
            PieceType::Queen => 9,
            PieceType::King => 50,
        }
    }

    pub fn from(letter: char) -> Result<Self, ()> {
        match letter.to_ascii_lowercase() {
            'p' => Ok(Self::Pawn(0)),
            'n' => Ok(Self::Knight),
            'b' => Ok(Self::Bishop),
            'r' => Ok(Self::Rook),
            'q' => Ok(Self::Queen),
            'k' => Ok(Self::King),
            _ => Err(()),
        }
    }

    pub fn letter(&self) -> char {
        match self {
            PieceType::Pawn(_) => 'p',
            PieceType::Knight => 'n',
            PieceType::Bishop => 'b',
            PieceType::Rook => 'r',
            PieceType::Queen => 'q',
            PieceType::King => 'k',
        }
    }
}

///# Color
///Represents a color. Used for square colors and piece colors.
///
///# Public Methods
///
///## value(&self) -> i32
///returns 1 for White and -1 for Black
///
///## opposite(&self) -> Self
///If self is white, return black. If self is black, return white.
///
///## home_rank(&self) -> usize
///Returns the 0 or 7
#[derive(Eq, PartialEq, Clone, Default, Copy)]
pub enum Color {
    #[default]
    White,
    Black,
}

impl fmt::Display for Color {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        let word = match self {
            Self::White => "White",
            Self::Black => "Black",
        };
        write!(f, "{word}")
    }
}

impl fmt::Debug for Color {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{self}")
    }
}

impl Color {
    pub fn value(&self) -> i32 {
        match self {
            Color::White => 1,
            Color::Black => -1,
        }
    }
    
    pub fn home_rank(&self) -> usize {
        match self {
            Color::White => 0,
            Color::Black => 7,
        }
    }

    pub fn opposite(&self) -> Self {
        match self {
            Color::White => Color::Black,
            Color::Black => Color::White,
        }
    }

    fn letter(&self) -> char {
        match self {
            Color::White => 'w',
            Color::Black => 'b',
        }
    }
}

///# Coordinate
///Represents the location of a square or piece. All fields are public.
///
///# Public Methods
///
///## is_valid(&self) -> bool
///Determines whether coordinate fits into an 8 by 8 board
///
///## with_offset(&self, rise: i32, run: i32) -> Coordinate
///Creates a new coordinate a certain number of rows and columns away from self.
///
///## from(repr: String) -> Coordinate 
///Returns a coordinate from a 2-character string of <col><row>
///This is not 0-indexed, so row is decremented by one.
#[derive(Clone, Copy, Default, PartialEq, Eq, Hash)]
pub struct Coordinate {
    pub(crate) row: usize,
    pub(crate) col: usize,
}

impl fmt::Debug for Coordinate {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "Coordinate col = {} row = {}", self.col, self.row)
    }
}

impl fmt::Display for Coordinate {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{}{}", LETTERS.chars().nth(self.col).unwrap_or('a'), self.row + 1)
    }
}

impl Coordinate {
    pub fn is_valid(&self) -> bool {
        self.row < 8 && self.col < 8
    }

    pub fn with_offset(&self, rise: i32, run: i32) -> Result<Coordinate, ()> {
        let mut new = *self;

        if rise < 0 {
            let offset = -rise as usize;

            //can't leave board
            if offset > new.row {
                return Err(());
            }

            new.row -= offset; //convert to positive number and subtract
        } else {
            new.row += rise as usize;
        }
        if run < 0 {
            let offset = -run as usize;

            //can't leave board
            if offset > new.col {
                return Err(());
            }
            new.col -= offset; //convert to positive number and subtract
        } else {
            new.col += run as usize;
        }

        if new.is_valid() { Ok(new) } else { Err(()) }
    }

    fn color(&self) -> Option<Color> {
        //if the row num plus the col num is even, return black. Else, white.
        if (self.col + self.row).is_multiple_of(2) {
            Some(Color::Black)
        } else {
            Some(Color::White)
        }
    }

    pub fn new(col: usize, row: usize) -> Self {
        Coordinate { row, col }
    }

    pub fn from(repr: &mut String) -> Self {
        let col = col_num(repr.remove(0));
        let row = repr.remove(0).to_string().parse::<usize>();
        
        Self {
            col,
            row: row.unwrap() - 1,
        }
    }
}

///# Square
///One square on the board.
///
///# Public Methods
///
///## value(&self) -> i32
///The value of the piece on the square. If there is no piece, value is 0.
#[derive(Debug, Copy, Clone, Default, Eq, PartialEq)]
pub struct Square {
    pub(crate) location: Coordinate,
    pub(crate) color: Color,
    pub(crate) piece: Option<Piece>,
}

impl Square {
    pub fn value(&self) -> i32 {
        if let Some(piece) = self.piece {
            piece.ptype.value()
        } else {
            0
        }
    }

    //dependency of Board::draw()
    fn draw(&self) -> String {
        //find ansi codes
        let bg = if self.color == Color::Black {
            "\x1b[40m"
        } else {
            "\x1b[47m"
        };
        let fg = if self.color == Color::Black {
            "\x1b[0;37m"
        } else {
            "\x1b[30m"
        };
        let piece = if let Some(p) = self.piece {
            p.draw()
        } else {
            String::from("  ")
        };

        format!("{bg}{fg}{piece}\x1b[0m")
    }
}

///# Piece
///A piece on a square.
///
///# Public Methods
///
///## potential_moves(&self, game: &board::Board) -> Vec<Action>
///Gets potential moves that don't collide, but doesn't check for legality
#[derive(Debug, Clone, Copy, Eq, PartialEq)]
pub struct Piece {
    pub(crate) color: Color,
    pub(crate) ptype: PieceType,
    pub(crate) location: Coordinate,
    pub(crate) has_moved: bool,
}

impl Piece {
    //dependency of Square::draw()
    fn draw(&self) -> String {
        format!("{}{}", self.color.letter(), self.ptype.letter())
    }
}

///# Board
///The board the game is played on.
///
///# Public fields
///
///## turn: Color
///The player whose turn it is.
///
///# Public Methods
///
///## new() -> Board
///Creates a new board.
///
///## populate_starting_pos(&mut self)
///
///## square(&self, col: char, row: usize) -> &Square<'_>
///Returns an immutable reference to the square at the specified locationl.
///
///## mut_square(&mut self, col: char, row: usize) -> &mut Square<'_>
///Like `square()` but returns mutable reference.
///
///## put_piece_on(&'a mut self, col: char, row: usize, piece: Piece<'a>)
///Consumes `piece` and puts it on the specified square.
///
///## pieces(&self) -> Vec<Piece>
///Iterates through all squares on board to find pieces.
///Use sparingly, because it runs in O(n) time.
///
///## white_pieces(&self) -> Vec<Piece>
///Uses pieces() but filters for white pieces.
///
///## black_pieces(&self) -> Vec<Piece>
///Uses pieces() but filters for black pieces.
///
///## is_check(&self, player: Color) -> bool
///Is player's king threatened?
///
///## is_checkmate(&self, player: Color) -> bool
///Is check and no legal moves?
///
///## remove_piece_on(&mut self, coord: &Coordinate)
///Deletes the piece at the coordinate and removes coordinate from piece locations list.
///
///## move_from(&mut self, from: &Coordinate, to: &Coordinate)
///Moves the piece on from to the square at to.
///Locations are managed automatically.
///
///## draw(&self) -> String
///Draws a crude ascii board to represent the current position.
///The fmt::Display implementation uses draw().
///
///## update(&mut self)
///Replaces move_info field with up-to-date one.
///
///## duplicate(&self) -> Self
///Creates an identical new Board.
///Runs in O(n) time, so use sparingly.
///
///## round_next_turn(&self) -> i32
///Gets what round it will be next turn
#[derive(Debug)]
pub struct Board {
    squares: [[Square; 8]; 8],
    pub(crate) locations: HashSet<Coordinate>, //stores locations of pieces
    pub(crate) turn: Color,
    pub(crate) round: i32, //the number of the set of 2 moves
    pub(crate) move_info: MoveInfo,
}

impl fmt::Display for Board {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{}", self.draw())
    }
}

impl Default for Board {
    fn default() -> Self {
        let mut new = Self::new();
        new.populate_starting_pos();
        new
    }
}

impl PartialEq for Board {
    //VERY INEFFICENT
    fn eq(&self, other: &Self) -> bool {
        if self.turn != other.turn || self.squares != other.squares {
            return false;
        } else {
            //search locations
            for loc in self.locations.clone() {
                if !other.locations.contains(&loc) {
                    return false;
                }
            }
        }

        true
    }
}

impl Board {
    pub fn new() -> Self {
        let mut uninit_arr: [MaybeUninit<[Square; 8]>; 8] = [MaybeUninit::uninit(); 8];

        for (col, col_arr_mem) in uninit_arr.iter_mut().enumerate() {
            //build vec of squares and convert to array
            let mut uninit_col_array: [MaybeUninit<Square>; 8] = [MaybeUninit::uninit(); 8];

            for (row, cell) in uninit_col_array.iter_mut().enumerate() {
                let coord = Coordinate::new(col, row);
                //get square color
                let color = coord.color().unwrap_or(Color::White);

                //create square
                cell.write(Square {
                    location: coord,
                    color,
                    piece: None,
                });
            }

            unsafe {
                let col_array =
                    mem::transmute::<[MaybeUninit<Square>; 8], [Square; 8]>(uninit_col_array);
                col_arr_mem.write(col_array);
            }
        }

        //convert uninitialized cols into array
        let grid;
        unsafe {
            grid = mem::transmute::<[MaybeUninit<[Square; 8]>; 8], [[Square; 8]; 8]>(uninit_arr);
        }

        Board {
            squares: grid,
            locations: HashSet::new(),
            turn: Color::White,
            round: 1, //not 0-indexed; 0 is used to signal "never" or a turn that is impossible
            move_info: MoveInfo::new(),
        }
    }

    pub fn populate_starting_pos(&mut self) {
        //go through each column and put corresponding piece
        self.populate_column(0, PieceType::Rook);
        self.populate_column(7, PieceType::Rook);
        self.populate_column(1, PieceType::Knight);
        self.populate_column(6, PieceType::Knight);
        self.populate_column(2, PieceType::Bishop);
        self.populate_column(5, PieceType::Bishop);
        self.populate_column(3, PieceType::Queen);
        self.populate_column(4, PieceType::King);
    }

    //this is private because it's just a dependency of populate_starting_pos()
    fn populate_column(&mut self, col: usize, ptype: PieceType) {
        //place non-pawns
        let c = Coordinate::new(col, 0);
        self.put_piece_on(
            &c,
            Piece {
                color: Color::White,
                ptype,
                location: c,
                has_moved: false,
            },
        );
        let c = Coordinate::new(col, 7);
        self.put_piece_on(
            &c,
            Piece {
                color: Color::Black,
                ptype,
                location: c,
                has_moved: false,
            },
        );

        //pawns
        let c = Coordinate::new(col, 1);
        self.put_piece_on(
            &c,
            Piece {
                color: Color::White,
                ptype: PieceType::Pawn(0),
                location: c,
                has_moved: false,
            },
        );
        let c = Coordinate::new(col, 6);
        self.put_piece_on(
            &c,
            Piece {
                color: Color::Black,
                ptype: PieceType::Pawn(0),
                location: c,
                has_moved: false,
            },
        );
    }

    pub fn square(&self, coord: &Coordinate) -> &Square {
        &self.squares[coord.col][coord.row]
    }

    pub fn mut_square(&mut self, coord: &Coordinate) -> &mut Square {
        &mut self.squares[coord.col][coord.row]
    }

    pub fn put_piece_on(&mut self, coord: &Coordinate, mut piece: Piece) {
        let s = self.mut_square(coord);
        piece.location = *coord;
        s.piece = Some(piece);
        self.locations.insert(*coord);
    }

    pub fn remove_on(&mut self, coord: &Coordinate) {
        //delete piece and remove location from locations
        self.mut_square(coord).piece = None;
        self.locations.remove(coord);
    }

    pub fn move_from(&mut self, from: &Coordinate, to: &Coordinate) {
        let from_square = self.mut_square(from);
        let mut moving_piece = from_square.piece.unwrap_or_else(|| {
            panic!("Moving from {from} to {to} isn't possible because there is no piece to move.")
        });
        moving_piece.has_moved = true;

        self.remove_on(from);
        self.put_piece_on(to, moving_piece);
    }

    ///Keep in mind that `pieces()` returns *copies* of pieces on the board, not the pieces
    ///themselves.
    pub fn pieces(&self) -> Vec<Piece> {
        let mut pieces = Vec::new();

        for loc in self.locations.iter() {
            if let Some(piece) = self.square(loc).piece {
                pieces.push(piece)
            } else {
                eprintln!(
                    "Board::pieces(): Location {loc} is in Board.locations, but there is no piece there!"
                );
            }
        }

        pieces
    }

    pub fn white_pieces(&self) -> Vec<Piece> {
        let mut pieces = self.pieces();
        pieces.retain(|p| p.color == Color::White);
        pieces
    }

    pub fn black_pieces(&self) -> Vec<Piece> {
        let mut pieces = self.pieces();
        pieces.retain(|p| p.color == Color::Black);
        pieces
    }

    pub fn draw(&self) -> String {
        let mut output = String::from("     a    b    c    d    e    f    g    h");

        //don't use for loop because decreasing ranges are hard
        let mut row: i32 = 7;
        while row >= 0 {
            //row number
            output.push_str(format!("\n\n{}", row + 1).as_str());

            for col in LETTERS.chars() {
                output.push_str(" | ");
                //get square, draw, convert to str
                output.push_str(
                    self.square(&Coordinate::new(col_num(col), row as usize))
                        .draw()
                        .as_str(),
                );
            }

            row -= 1;
        }

        output
    }

    pub fn duplicate(&self) -> Self {
        let mut new = Self::new();

        //iterate over pieces and place on new board
        for piece in self.pieces() {
            new.put_piece_on(&piece.location, piece);
        }

        //set other fields
        new.turn = self.turn;
        new.round = self.round;
        new.update();

        new
    }

    pub fn round_next_turn(&self) -> i32 {
        match self.turn {
            Color::White => self.round,
            Color::Black => self.round + 1,
        }
    }
}

pub fn col_num(name: char) -> usize {
    let res = LETTERS.find(name);
    if let Some(num) = res {
        num
    } else {
        panic!("Column {name} doesn't exist!");
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn default_board() -> Board {
        let mut b = Board::new();
        b.populate_starting_pos();
        b
    }

    #[test]
    fn positive_offset() {
        let c = Coordinate::new(5, 5);
        let new = c.with_offset(1, 1);
        assert_eq!(new, Ok(Coordinate::new(6, 6)));
    }

    #[test]
    fn negative_offset() {
        let c = Coordinate::new(5, 5);
        let new = c.with_offset(-5, -5);
        assert_eq!(new, Ok(Coordinate::new(0, 0)));
    }

    #[test]
    fn num_of_pieces() {
        let b = default_board();
        assert_eq!(b.pieces().len(), 32);
        assert_eq!(b.white_pieces().len(), 16);
        assert_eq!(b.black_pieces().len(), 16);
    }

    #[test]
    fn coordinate_from() {
        let expected = Coordinate::new(0, 0);
        let generated = Coordinate::from(&mut String::from("a1"));
        assert_eq!(generated, expected);
    }
}
